infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java

```diff
@@ -21,43 +23,76 @@
  */
 @Data
 @Builder
+@Slf4j
 public class PollReport {
 
     @Default
-    private long pollEpoch = Layout.INVALID_EPOCH;
+    private final long pollEpoch = Layout.INVALID_EPOCH;
+
     /**
-     * Contains list of servers successfully connected with current node.
-     * It doesn't contain nodes answered with WrongEpochException.
+     * A map of nodes answered with {@link WrongEpochException}.
      */
     @Default
-    private final ImmutableSet<String> connectedNodes = ImmutableSet.of();
+    private final ImmutableMap<String, Long> wrongEpochs = ImmutableMap.of();
+
     /**
-     * Contains list of servers disconnected from this node. If the node A can't ping node B then node B will be added
-     * to failedNodes list.
+     * Current cluster state, collected by a failure detector
      */
-    @Default
-    private final ImmutableSet<String> failedNodes = ImmutableSet.of();
+    @NonNull
+    private final ClusterState clusterState;
+
     /**
-     * A map of nodes answered with {@link WrongEpochException}.
+     * List of responsive servers in a current layout
      */
-    @Default
-    private final ImmutableMap<String, Long> wrongEpochs = ImmutableMap.of();
+    @NonNull
+    private final List<String> responsiveServers;
+
     /**
-     * Indicates if current layout slot is unfilled
+     * Returns all connected nodes to the current node including nodes with higher epoch.
+     *
+     * @return set of all reachable nodes
      */
-    @Default
-    private final boolean currentLayoutSlotUnFilled = false;
+    public ImmutableSet<String> getAllReachableNodes() {
+        return Sets.union(getReachableNodes(), wrongEpochs.keySet()).immutableCopy();
+    }
+
     /**
-     * Current cluster state, collected by a failure detector
+     * Contains list of servers successfully connected with current node.
+     * It doesn't contain nodes answered with WrongEpochException.
      */
-    @NonNull
-    private final ClusterState clusterState;
+    public Set<String> getReachableNodes() {
+        Set<String> connectedNodes =  clusterState.getLocalNodeConnectivity().getConnectedNodes();
+        return Sets.difference(connectedNodes, wrongEpochs.keySet());
+    }
 
     /**
-     * Returns all connected nodes to the current node
-     * @return
+     * Contains list of servers disconnected from this node. If the node A can't ping node B then node B will be added
+     * to failedNodes list.
      */
-    public ImmutableSet<String> getAllConnectedNodes() {
-        return Sets.union(connectedNodes, wrongEpochs.keySet()).immutableCopy();
+    public Set<String> getFailedNodes() {
+        Set<String> failedNodes = clusterState.getLocalNodeConnectivity().getFailedNodes();
+
+        return Sets.difference(failedNodes, wrongEpochs.keySet());
```

Comment:
Isn't the set difference always empty? When you get wrongEpoch does this mean the ping has not succeeded?

Reply:
the ping has succeeded https://github.com/CorfuDB/CorfuDB/pull/1831/files#diff-45a8e0bb947ac9cb433401fb4d2d24e9R113